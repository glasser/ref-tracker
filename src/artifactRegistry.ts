import * as core from '@actions/core';
import { ArtifactRegistryClient } from '@google-cloud/artifact-registry';

export interface GetAllEquivalentTagsOptions {
  /** The name of the specific Docker image in question (ie, a Docker
   * "repository", not an Artifact Registry "repository" that contains them.) */
  dockerImageRepository: string;
  tag: string;
}

export interface DockerRegistryClient {
  getAllEquivalentTags(options: GetAllEquivalentTagsOptions): Promise<string[]>;
}

export class ArtifactRegistryDockerRegistryClient {
  private client: ArtifactRegistryClient;
  constructor(
    /** A string of the form
     * `projects/PROJECT/locations/LOCATION/repositories/REPOSITORY`; this is an
     * Artifact Registry repository, which is a set of Docker repositories. */
    private artifactRegistryRepository: string,
  ) {
    this.client = new ArtifactRegistryClient();
  }

  async getAllEquivalentTags({
    dockerImageRepository,
    tag,
  }: GetAllEquivalentTagsOptions): Promise<string[]> {
    // Input is relatively trusted; this is largely to prevent mistaken uses
    // like specifying a full Docker-style repository with slashes.
    if (dockerImageRepository.includes('/')) {
      throw Error('repository cannot contain a slash');
    }
    if (tag.includes('/')) {
      throw Error('tag cannot contain a slash');
    }

    const tagPrefix = `${this.artifactRegistryRepository}/packages/${dockerImageRepository}/tags/`;

    core.info(`Fetching tag ${tagPrefix}${tag}`);
    // Note: this throws if the repository or tag are not found.
    const { version } = (
      await this.client.getTag({ name: `${tagPrefix}${tag}` })
    )[0];
    if (!version) {
      throw Error(`No version found for ${dockerImageRepository}@${tag}`);
    }

    core.info(`Fetching version ${version}`);
    const { relatedTags } = (
      await this.client.getVersion({ name: version, view: 'FULL' })
    )[0];
    if (!relatedTags) {
      throw Error(`No related tags returned for ${version}`);
    }
    return relatedTags.map(({ name }) => {
      if (!name?.startsWith(tagPrefix)) {
        throw Error(
          `Expected tag name to start with '${tagPrefix}'; got '${name}'`,
        );
      }
      return name.slice(tagPrefix.length);
    });
  }
}
