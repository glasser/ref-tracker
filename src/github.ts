import { getOctokit } from '@actions/github';
import * as core from '@actions/core';

export interface ResolveRefToShaOptions {
  repoURL: string;
  ref: string;
}

export interface GetTreeSHAForPathOptions {
  repoURL: string;
  ref: string;
  path: string;
}

export interface GitHubClient {
  resolveRefToSha(options: ResolveRefToShaOptions): Promise<string>;
  getTreeSHAForPath(options: GetTreeSHAForPathOptions): Promise<string>;
}

interface OwnerAndRepo {
  owner: string;
  repo: string;
}

function parseRepoURL(repoURL: string): OwnerAndRepo {
  const m = repoURL.match(/\bgithub\.com\/([\w.-]+)\/([\w.-]+?)(?:\.git|\/)?$/);
  if (!m) {
    throw Error(`Can only track GitHub repoURLs, not ${repoURL}`);
  }
  return { owner: m[1], repo: m[2] };
}

export class OctokitGitHubClient {
  constructor(private octokit: ReturnType<typeof getOctokit>) {}

  async resolveRefToSha({
    repoURL,
    ref,
  }: ResolveRefToShaOptions): Promise<string> {
    const { owner, repo } = parseRepoURL(repoURL);
    const sha = (
      await this.octokit.rest.repos.getCommit({
        owner,
        repo,
        ref,
        mediaType: {
          format: 'sha',
        },
      })
    ).data as unknown;
    // The TS types don't understand that `mediaType: {format: 'sha'}` turns
    // `.data` into a string, so we have to cast to `unknown` and check
    // ourselves.
    if (typeof sha !== 'string') {
      throw Error('Expected string response');
    }
    return sha;
  }

  async getTreeSHAForPath({
    repoURL,
    ref,
    path,
  }: GetTreeSHAForPathOptions): Promise<string> {
    const { owner, repo } = parseRepoURL(repoURL);
    // FIXME error handling
    const content = await this.octokit.rest.repos.getContent({
      owner,
      repo,
      ref,
      path,
      mediaType: {
        format: 'object',
      },
    });
    core.info(`Got content ${JSON.stringify(content, null, 2)}`);
    return 'yay';
  }
}
