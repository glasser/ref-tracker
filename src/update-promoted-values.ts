import * as core from '@actions/core';
import { RE2 } from 're2-wasm';
import * as yaml from 'yaml';
import { ScalarTokenWriter, getTopLevelBlocks } from './yaml';

// FIXME we will want to make a way to filter promote-ables so that we can
// separate "promote to staging" from "promote to dev" for pipelines that have
// multiple

interface Promote {
  scalarTokenWriter: ScalarTokenWriter;
  value: string;
}

export async function updatePromotedValues(
  contents: string,
  promotionTargetRegexp: string | null,
): Promise<string> {
  return core.group('Processing promote', async () => {
    // We use re2-wasm instead of built-in RegExp so we don't have to worry about
    // REDOS attacks.
    const promotionTargetRE2 = promotionTargetRegexp
      ? new RE2(promotionTargetRegexp, 'u')
      : null;

    // See updateGitRefs for an explanation about our approach to parsing YAML.
    core.info('Parsing');
    const topLevelTokens = [...new yaml.Parser().parse(contents)];
    const documents = [
      ...new yaml.Composer({ keepSourceTokens: true }).compose(topLevelTokens),
    ];

    // We decide what to do and then we do it, just in case there are any
    // overlaps between our reads and writes.
    core.info('Looking for promote');
    const promotes = documents.flatMap((document) =>
      findPromotes(document, promotionTargetRE2),
    );

    core.info('Copying values');
    for (const { scalarTokenWriter, value } of promotes) {
      scalarTokenWriter.write(value);
    }
    core.info('Stringifying');
    return topLevelTokens
      .map((topLevelToken) => yaml.CST.stringify(topLevelToken))
      .join('');
  });
}

function findPromotes(
  document: yaml.Document.Parsed,
  promotionTargetRE2: RE2 | null,
): Promote[] {
  const { blocks } = getTopLevelBlocks(document);
  const promotes: Promote[] = [];
  for (const [myName, me] of blocks) {
    if (promotionTargetRE2 && !promotionTargetRE2.test(myName)) {
      continue;
    }
    if (!me.has('promote')) {
      continue;
    }
    const promote = me.get('promote');
    if (!yaml.isMap(promote)) {
      throw Error(`The value at ${myName}.promote must be a map`);
    }
    const from = promote.get('from');
    if (typeof from !== 'string') {
      throw Error(`The value at ${myName}.promote.from must be a string`);
    }
    const fromBlock = blocks.get(from);
    if (!fromBlock) {
      throw Error(
        `The value at ${myName}.promote.from must reference a top-level key with map value`,
      );
    }

    // FIXME use default yamlPaths
    const yamlPathsSeq = promote.get('yamlPaths');
    if (!yaml.isSeq(yamlPathsSeq)) {
      throw Error(`The value at ${myName}.promote.yamlPaths must be an array`);
    }
    const yamlPaths = yamlPathsSeq.toJSON();
    if (!Array.isArray(yamlPaths)) {
      throw Error('YAMLSeq.toJSON surprisingly did not return an array');
    }
    if (!yamlPaths.every(isCollectionPath)) {
      throw Error(
        `The value at ${myName}.promote.yamlPaths must be an array whose elements are arrays of strings or numbers`,
      );
    }

    for (const collectionPath of yamlPaths) {
      const sourceValue = fromBlock.getIn(collectionPath);
      if (typeof sourceValue !== 'string') {
        throw Error(`Could not promote from ${[from, ...collectionPath]}`);
      }
      // true means keepScalar, ie get the scalar node to write.
      const targetNode = me.getIn(collectionPath, true);
      if (!yaml.isScalar(targetNode)) {
        throw Error(`Could not promote to ${[myName, ...collectionPath]}`);
      }
      const scalarToken = targetNode.srcToken;
      if (!yaml.CST.isScalar(scalarToken)) {
        // this probably can't happen, but let's make the types happy
        throw Error(
          `${[myName, ...collectionPath]} value must come from a scalar token`,
        );
      }
      promotes.push({
        scalarTokenWriter: new ScalarTokenWriter(scalarToken, document.schema),
        value: sourceValue,
      });
    }
  }
  return promotes;
}

type CollectionPath = CollectionIndex[];
type CollectionIndex = string | number;

function isCollectionPath(value: unknown): value is CollectionPath {
  return Array.isArray(value) && value.every(isCollectionIndex);
}

function isCollectionIndex(value: unknown): value is CollectionIndex {
  return typeof value === 'string' || typeof value === 'number';
}
