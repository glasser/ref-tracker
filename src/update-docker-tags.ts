import * as core from '@actions/core';
import min from 'lodash/min';
import * as yaml from 'yaml';
import { DockerRegistryClient } from './artifactRegistry';
import {
  ScalarTokenWriter,
  getStringAndScalarTokenFromMap,
  getStringValue,
  getTopLevelBlocks,
} from './yaml';

interface Trackable {
  trackMutableTag: string;
  dockerImageRepository: string;
  tag: string;
  tagScalarTokenWriter: ScalarTokenWriter;
}

export async function updateDockerTags(
  contents: string,
  dockerRegistryClient: DockerRegistryClient,
): Promise<string> {
  return core.group('Processing trackMutableTag', async () => {
    // See updateGitRefs for an explanation about our approach to parsing YAML.
    core.info('Parsing');
    const topLevelTokens = [...new yaml.Parser().parse(contents)];
    const documents = [
      ...new yaml.Composer({ keepSourceTokens: true }).compose(topLevelTokens),
    ];

    if (documents.length > 1) {
      throw new Error('Multiple documents in YAML file');
    }

    if (documents.length < 1) {
      return contents;
    }

    const document = documents[0];

    core.info('Looking for trackMutableTag');
    const trackables = findTrackables(document);

    core.info('Checking tags against Artifact Registry');
    await checkTagsAgainstArtifactRegistryAndModifyScalars(
      trackables,
      dockerRegistryClient,
    );
    core.info('Stringifying');
    return topLevelTokens
      .map((topLevelToken) => yaml.CST.stringify(topLevelToken))
      .join('');
  });
}

function findTrackables(doc: yaml.Document.Parsed): Trackable[] {
  const trackables: Trackable[] = [];

  const { blocks, globalBlock } = getTopLevelBlocks(doc);

  let globalDockerImageRepository: string | null = null;

  if (globalBlock?.has('dockerImage')) {
    const dockerImageBlock = globalBlock.get('dockerImage');
    if (!yaml.isMap(dockerImageBlock)) {
      throw Error('Document has `global.dockerImageBlock` that is not a map');
    }
    // Read repository from 'global' (keeping it null if it's not
    // there, though throwing if it's there as non-strings).
    globalDockerImageRepository = getStringValue(
      dockerImageBlock,
      'repository',
    );
  }

  for (const [key, value] of blocks) {
    if (!value.has('dockerImage')) {
      continue;
    }
    const dockerImageBlock = value.get('dockerImage');
    if (!yaml.isMap(dockerImageBlock)) {
      throw Error(`Document has \`${key}.dockerImage\` that is not a map`);
    }

    const dockerImageRepository =
      getStringValue(dockerImageBlock, 'repository') ??
      globalDockerImageRepository;
    const trackMutableTag = getStringValue(dockerImageBlock, 'trackMutableTag');
    const tagScalarTokenAndValue = getStringAndScalarTokenFromMap(
      dockerImageBlock,
      'tag',
    );

    if (trackMutableTag && dockerImageRepository && tagScalarTokenAndValue) {
      trackables.push({
        trackMutableTag,
        dockerImageRepository,
        tag: tagScalarTokenAndValue.value,
        tagScalarTokenWriter: new ScalarTokenWriter(
          tagScalarTokenAndValue.scalarToken,
          doc.schema,
        ),
      });
    }
  }

  return trackables;
}

async function checkTagsAgainstArtifactRegistryAndModifyScalars(
  trackables: Trackable[],
  dockerRegistryClient: DockerRegistryClient,
): Promise<void> {
  for (const trackable of trackables) {
    const prefix = `${trackable.trackMutableTag}---`;

    const equivalentTags = (
      await dockerRegistryClient.getAllEquivalentTags({
        dockerImageRepository: trackable.dockerImageRepository,
        tag: trackable.trackMutableTag,
      })
    ).filter((t) => t.startsWith(prefix));

    // We assume that all the tags with the triple-dash in them are immutable:
    // once they point at a particular SHA, they never change. (Whereas the tag
    // we're "tracking" is mutable.)
    //
    // If the current tag has the right format for an immutable tag and it
    // points to the same image as the mutable tag, leave it alone: there's no
    // reason to create a no-op diff.
    if (equivalentTags.includes(trackable.tag)) {
      core.info(
        `for image ${trackable.dockerImageRepository}:${trackable.trackMutableTag}, preserving current tag ${trackable.tag}`,
      );
      continue;
    }
    // We can choose *any* of these equivalent triple-dashed tags, and it will
    // select the correct image version.
    //
    // Our tag structure increase over time (by including the number of commits
    // since the start as determined by `git rev-list --first-parent --count
    // HEAD`), and also includes the git commit at which it was built.
    //
    // So by choosing the lexicographically earliest of the equivalent tags, we
    // are most likely to choose a tag where the named git commit actually is
    // the commit that made a relevant change that affected the image.
    //
    // Additionally, this means that reverting a code change is likely to result
    // in a revert of the tag.  Imagine that tag `main---00123-abcd` is
    // currently running in both staging and prod, and a code change moves
    // `main` to `main---00130-bcde` and this is deployed to staging (opening a
    // prod promotion PR). A bug is found, so the code change X is reverted.
    // Reproducible builds will mean that the newest tag `main---00134-dcba`
    // will hopefully point to the same image version as `main---00123-abcd`.
    // Choosing the min version here will mean that we will in fact "revert"
    // staging to `main---00123-abcd`. This is now the exact same tag that is
    // running in prod, so the prod promotion PR can auto-close rather than
    // encouraging us to consider a no-op deploy to prod.
    const earliestMatchingTag = min(equivalentTags);
    if (!earliestMatchingTag) {
      throw new Error(
        `No tags on ${trackable.dockerImageRepository} start with '${prefix}'`,
      );
    }

    // It's OK if the current one is null because that's what we're overwriting, but we shouldn't
    // overwrite *to* something that doesn't exist.
    core.info(
      `for image ${trackable.dockerImageRepository}:${trackable.trackMutableTag}, changing to minimal matching tag ${earliestMatchingTag}`,
    );
    trackable.tagScalarTokenWriter.write(earliestMatchingTag);
  }
}
