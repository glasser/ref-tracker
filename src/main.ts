import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import * as yaml from 'yaml';
import { readFile, writeFile } from 'fs/promises';
import { GitHubClient, OctokitGitHubClient } from './github';

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function run(): Promise<void> {
  try {
    const githubToken = core.getInput('github-token');
    const octokit = github.getOctokit(githubToken);
    // FIXME consider adding @octokit/plugin-throttling
    const gitHubClient = new OctokitGitHubClient(octokit);
    const files = core.getInput('files');
    const globber = await glob.create(files);
    const filenames = await globber.glob();
    for (const filename of filenames) {
      core.info(`Looking at ${filename}`);
      await processFile(filename, gitHubClient);
    }
  } catch (error) {
    // Fail the workflow run if an error occurs
    if (error instanceof Error) core.setFailed(error.message);
  }
}

interface Trackable {
  trackMutableRef: string;
  repoURL: string;
  path: string;
  ref: string;
  newRef: string;
  startCharacter: number;
  endCharacter: number;
}

export async function processFile(
  filename: string,
  gitHubClient: GitHubClient,
): Promise<void> {
  const newContents = await newContentsForFile(filename, gitHubClient);
  if (newContents !== null) {
    await writeFile(filename, newContents);
  }
}

export function parseDocument(
  contents: string,
  filename: string,
): yaml.Document | null {
  const doc = yaml.parseDocument(contents);
  if (doc.errors.length) {
    core.error(`Errors parsing YAML file ${filename}; skipping processing`);
    for (const error of doc.errors) {
      core.error(error.message);
    }
    return null;
  }
  if (doc.warnings.length) {
    core.warning(`Warnings parsing YAML file ${filename}`);
    for (const warning of doc.warnings) {
      core.warning(warning.message);
    }
  }
  return doc;
}

export function findTrackables(doc: yaml.Document): Trackable[] {
  const trackables: Trackable[] = [];

  // FIXME figure out error handling

  yaml.visit(doc, {
    Map(_, node) {
      if (
        node.has('trackMutableRef') &&
        node.has('repoURL') &&
        node.has('path') &&
        node.has('ref')
      ) {
        const trackMutableRef = node.get('trackMutableRef');
        if (typeof trackMutableRef !== 'string') {
          throw Error(`trackMutableRef value must be a string`);
        }
        const repoURL = node.get('repoURL');
        if (typeof repoURL !== 'string') {
          throw Error(`repoURL value must be a string`);
        }
        const path = node.get('path');
        if (typeof path !== 'string') {
          throw Error(`path value must be a string`);
        }
        const refScalar = node.get('ref', true);
        if (!yaml.isScalar(refScalar)) {
          throw Error('ref value must be a scalar');
        }
        const ref = refScalar.value;
        if (typeof ref !== 'string') {
          throw Error('ref value must be a string');
        }
        if (!refScalar.range) {
          // Shouldn't happen for a Scalar created by the parser.
          throw Error('YAML for ref does not say where it lives');
        }
        const [startCharacter, endCharacter] = refScalar.range;
        trackables.push({
          trackMutableRef,
          repoURL,
          path,
          ref,
          newRef: ref,
          startCharacter,
          endCharacter,
        });
      }
    },
  });

  // Just in case we don't produce them in order, make them be in order now.
  trackables.sort((a, b) => a.startCharacter - b.startCharacter);

  return trackables;
}

export async function newContentsForFile(
  filename: string,
  gitHubClient: GitHubClient,
): Promise<string | null> {
  core.info('reading');
  const contents = await readFile(filename, 'utf-8');
  core.info('parsing');
  const doc = parseDocument(contents, filename);
  if (!doc) {
    return null;
  }

  core.info('finding trackables');
  const trackables = findTrackables(doc);
  core.info('updating refs');
  await updateRefsFromGitHub(trackables, gitHubClient);
  core.info('rewriting refs');
  return rewriteRefs(contents, trackables);
}

export async function updateRefsFromGitHub(
  trackables: Trackable[],
  gitHubClient: GitHubClient,
): Promise<void> {
  for (const trackable of trackables) {
    // FIXME error handling
    trackable.newRef = await gitHubClient.resolveRefToSha({
      repoURL: trackable.repoURL,
      ref: trackable.ref,
    });
  }
}

export function rewriteRefs(contents: string, trackables: Trackable[]): string {
  // First, let's validate that the substrings we're trying to update come in
  // order and don't overlap.
  for (const [i, trackable] of trackables.entries()) {
    if (trackable.endCharacter <= trackable.startCharacter) {
      throw Error(
        `Trackable has end ${trackable.endCharacter} that is not after start ${trackable.startCharacter}`,
      );
    }
    if (i > 0 && trackable.startCharacter < trackables[i - 1].endCharacter) {
      throw Error(
        `Trackable has start ${
          trackable.startCharacter
        } that is before previous end ${trackables[i - 1].endCharacter}`,
      );
    }
  }

  let adjustment = 0;
  for (const trackable of trackables) {
    // Wrap quotes around the ref (needed if the ref is all digits).
    const newValue = JSON.stringify(trackable.newRef);
    contents =
      contents.substring(0, trackable.startCharacter + adjustment) +
      newValue +
      contents.substring(trackable.endCharacter + adjustment);

    // If we increased the length of the part we changed, we'll need to increase
    // all the indexes we use later. If we decrease the length, we'll need to
    // decrease the indexes.
    adjustment +=
      newValue.length - (trackable.endCharacter - trackable.startCharacter);
  }

  return contents;
}
