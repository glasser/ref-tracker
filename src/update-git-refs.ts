import * as core from '@actions/core';
import * as yaml from 'yaml';
import { GitHubClient } from './github';
import {
  ScalarTokenWriter,
  getStringAndScalarTokenFromMap,
  getStringValue,
  getTopLevelBlocks,
} from './yaml';

interface Trackable {
  trackMutableRef: string;
  repoURL: string;
  path: string;
  ref: string;
  refScalarTokenWriter: ScalarTokenWriter;
}

export async function updateGitRefs(
  contents: string,
  gitHubClient: GitHubClient,
): Promise<string> {
  return core.group('Processing trackMutableRef', async () => {
    // The yaml module lets us parse YAML into three layers of abstraction:
    // - It can create raw JS arrays/objects/etc, which is simple to dealing
    //   with but loses track of everything relating to formatting.
    // - It can create a low-level "Concrete Syntax Tree" (CST) which lets us
    //   re-create the original document with byte-by-byte accuracy, but is
    //   awkward to read from (eg, you have to navigate maps item by item rather
    //   than using keys).
    // - It can create a high-level "Abstract Syntax Tree" (AST) which is easier
    //   to read from but loses some formatting details.
    //
    // We'd prefer to read ASTs and write CSTs, and in fact the module lets us
    // do exactly that. We first create CSTs with the "Parser". We then convert
    // it into ASTs with the Composer, passing in the `keepSourceTokens` option
    // which means that every node in the AST will have a `srcToken` reference
    // to the underlying CST node that created it. When we want to make changes,
    // we do that by writing to the CST node found in a `srcToken` reference.
    // Finally, when we're done, we stringify the CSTs (which have been mutated)
    // rather than the ASTs.
    core.info('Parsing');
    const topLevelTokens = [...new yaml.Parser().parse(contents)];
    const documents = [
      ...new yaml.Composer({ keepSourceTokens: true }).compose(topLevelTokens),
    ];

    // These files are all Helm values.yaml files, and Helm doesn't support a
    // multiple-document stream (with ---) for its value files. Or well, it
    // ignores any documents after the first, so there's no point in allowing
    // folks to put them in our codebase.
    if (documents.length > 1) {
      throw new Error('Multiple documents in YAML file');
    }

    // If the file is empty (or just whitespace or whatever), that's fine; we
    // can just leave it alone.
    if (documents.length < 1) {
      return contents;
    }

    const document = documents[0];

    core.info('Looking for trackMutableRef');
    const trackables = findTrackables(document);

    core.info('Checking refs against GitHub');
    await checkRefsAgainstGitHubAndModifyScalars(trackables, gitHubClient);
    core.info('Stringifying');
    return topLevelTokens
      .map((topLevelToken) => yaml.CST.stringify(topLevelToken))
      .join('');
  });
}

function findTrackables(doc: yaml.Document.Parsed): Trackable[] {
  const trackables: Trackable[] = [];

  // FIXME figure out error handling

  const { blocks, globalBlock } = getTopLevelBlocks(doc);

  let globalRepoURL: string | null = null;
  let globalPath: string | null = null;

  if (globalBlock?.has('gitConfig')) {
    const gitConfigBlock = globalBlock.get('gitConfig');
    if (!yaml.isMap(gitConfigBlock)) {
      throw Error('Document has `global.gitConfig` that is not a map');
    }
    // Read repoURL and path from 'global' (keeping them null if they're not
    // there, though throwing if they're there as non-strings).
    globalRepoURL = getStringValue(gitConfigBlock, 'repoURL');
    globalPath = getStringValue(gitConfigBlock, 'path');
  }

  for (const [key, value] of blocks) {
    if (!value.has('gitConfig')) {
      continue;
    }
    const gitConfigBlock = value.get('gitConfig');
    if (!yaml.isMap(gitConfigBlock)) {
      throw Error(`Document has \`${key}.gitConfig\` that is not a map`);
    }

    const repoURL = getStringValue(gitConfigBlock, 'repoURL') ?? globalRepoURL;
    const path = getStringValue(gitConfigBlock, 'path') ?? globalPath;
    const trackMutableRef = getStringValue(gitConfigBlock, 'trackMutableRef');
    const refScalarTokenAndValue = getStringAndScalarTokenFromMap(
      gitConfigBlock,
      'ref',
    );

    if (trackMutableRef && repoURL && path && refScalarTokenAndValue) {
      trackables.push({
        trackMutableRef,
        repoURL,
        path,
        ref: refScalarTokenAndValue.value,
        refScalarTokenWriter: new ScalarTokenWriter(
          refScalarTokenAndValue.scalarToken,
          doc.schema,
        ),
      });
    }
  }

  return trackables;
}

async function checkRefsAgainstGitHubAndModifyScalars(
  trackables: Trackable[],
  gitHubClient: GitHubClient,
): Promise<void> {
  for (const trackable of trackables) {
    // FIXME error handling
    const mutableRefCurrentSHA = await gitHubClient.resolveRefToSha({
      repoURL: trackable.repoURL,
      ref: trackable.trackMutableRef,
    });

    if (trackable.ref === trackable.trackMutableRef) {
      // The mutable ref was written down in ref too. We always want to replace
      // that with the SHA (and if we do the path-based check below we won't,
      // because they're the same). This is something that might happen when
      // you're first adding an app (ie just writing the same thing twice and
      // letting the automation "correct" it to a SHA).
      trackable.refScalarTokenWriter.write(mutableRefCurrentSHA);
      continue;
    }

    if (trackable.ref === mutableRefCurrentSHA) {
      // The thing we would write is already in the file.
      continue;
    }

    // OK, we've got a SHA that we could overwrite the current ref
    // (`trackable.ref`) with in the config file. But we don't want to do this
    // if it would be a no-op. Let's check the tree SHA
    // (https://git-scm.com/book/en/v2/Git-Internals-Git-Objects#_tree_objects)
    // at the given path to see if it has changed between `trackable.ref` and
    // the SHA we're thinking about replacing it with.
    const currentTreeSHA = await gitHubClient.getTreeSHAForPath({
      repoURL: trackable.repoURL,
      ref: trackable.ref,
      path: trackable.path,
    });
    const newTreeSHA = await gitHubClient.getTreeSHAForPath({
      repoURL: trackable.repoURL,
      ref: mutableRefCurrentSHA,
      path: trackable.path,
    });
    if (newTreeSHA === null) {
      throw Error(
        `Could not get tree SHA for ${mutableRefCurrentSHA} in ${trackable.repoURL} for ref ${trackable.path}`,
      );
    }
    // It's OK if the current one is null because that's what we're overwriting, but we shouldn't
    // overwrite *to* something that doesn't exist.
    core.info(
      `for path ${trackable.path}, got tree shas ${currentTreeSHA} for ${trackable.ref} and ${newTreeSHA} for ${mutableRefCurrentSHA}`,
    );
    if (currentTreeSHA === newTreeSHA) {
      core.info('(unchanged)');
    } else {
      core.info('(changed!)');
      trackable.refScalarTokenWriter.write(mutableRefCurrentSHA);
    }
  }
}
