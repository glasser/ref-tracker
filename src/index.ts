import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import { throttling } from '@octokit/plugin-throttling';
import { readFile, writeFile } from 'fs/promises';
import { ArtifactRegistryDockerRegistryClient } from './artifactRegistry';
import { OctokitGitHubClient } from './github';
import { updateDockerTags } from './update-docker-tags';
import { updateGitRefs } from './update-git-refs';
import { updatePromotedValues } from './update-promoted-values';

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function main(): Promise<void> {
  try {
    const files = core.getInput('files');
    const globber = await glob.create(files);
    const filenames = await globber.glob();
    for (const filename of filenames) {
      await processFile(filename);
    }
  } catch (error) {
    // Fail the workflow run if an error occurs
    if (error instanceof Error) core.setFailed(error.message);
  }
}

async function processFile(filename: string): Promise<void> {
  return core.group(`Processing ${filename}`, async () => {
    let contents = await readFile(filename, 'utf-8');

    if (core.getBooleanInput('update-git-refs')) {
      const githubToken = core.getInput('github-token');
      const octokit = github.getOctokit(githubToken, throttling);
      const gitHubClient = new OctokitGitHubClient(octokit);
      contents = await updateGitRefs(contents, gitHubClient);
    }

    const artifactRegistryRepository = core.getInput(
      'update-docker-tags-for-artifact-registry-repository',
    );

    if (artifactRegistryRepository) {
      const dockerRegistryClient = new ArtifactRegistryDockerRegistryClient(
        artifactRegistryRepository,
      );
      contents = await updateDockerTags(contents, dockerRegistryClient);
    }

    if (core.getBooleanInput('update-promoted-values')) {
      contents = await updatePromotedValues(
        contents,
        core.getInput('promotion-target-regexp') || null,
      );
    }

    await writeFile(filename, contents);
  });
}

main();
