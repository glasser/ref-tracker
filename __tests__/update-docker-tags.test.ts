import { readFile } from 'fs/promises';
import { join } from 'path';
import { GetAllEquivalentTagsOptions } from '../src/artifactRegistry';
import { updateDockerTags } from '../src/update-docker-tags';

// FIXME figure out logging (eg pass in a fake implementation of core
// to code, or mock it, or capture process.stdout, etc)
// Mock the GitHub Actions core library
// let infoMock: jest.SpyInstance;

async function fixture(filename: string): Promise<string> {
  return await readFile(
    join(__dirname, '__fixtures__', 'update-docker-tags', filename),
    'utf-8',
  );
}

describe('action', () => {
  it('updates docker tags', async () => {
    const contents = await fixture('sample.yaml');
    const dockerRegistryClient = {
      async getAllEquivalentTags({ tag }: GetAllEquivalentTagsOptions) {
        return (
          {
            'from-mutable': [
              'from-mutable---000123-abcd',
              'another',
              'from-mutable',
            ],
            'from-something-random': [
              'bla---000123-abcd',
              'from-something-random---000123-abcd',
            ],
            'needs-update': ['needs-update---000200-dbca'],
            'can-stay-same': [
              'can-stay-same---0123-abcd',
              'can-stay-same---0124-bbcd',
            ],
            'should-roll-back-not-forward': [
              'should-roll-back-not-forward---000100-abcd',
              'should-roll-back-not-forward---000200-dcba',
            ],
          }[tag] ?? []
        );
      },
    };
    const newContents = await updateDockerTags(contents, dockerRegistryClient);
    expect(newContents).toMatchSnapshot();

    // It should be idempotent in this case.
    expect(await updateDockerTags(contents, dockerRegistryClient)).toBe(
      newContents,
    );
  });
});
