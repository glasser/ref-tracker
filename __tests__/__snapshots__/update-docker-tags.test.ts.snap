// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action updates docker tags 1`] = `
"some-service-dev0:
  dockerImage:
    repository: foo
    trackMutableTag: from-mutable
    tag: from-mutable---000123-abcd

some-service-dev1:
  dockerImage:
    repository: foo
    trackMutableTag: from-something-random
    tag: from-something-random---000123-abcd

some-service-dev2:
  dockerImage:
    repository: foo
    trackMutableTag: needs-update
    tag: needs-update---000200-dbca


some-service-dev3:
  dockerImage:
    repository: foo
    trackMutableTag: can-stay-same
    tag: can-stay-same---0123-abcd

some-service-dev4:
  dockerImage:
    repository: foo
    trackMutableTag: should-roll-back-not-forward
    tag: should-roll-back-not-forward---000100-abcd
"
`;
